ui/components/app/simulation-details/useBalanceChanges.ts

```diff
@@ -0,0 +1,175 @@
+import { createSelector } from 'reselect';
+import {
+  SimulationBalanceChange,
+  SimulationData,
+  SimulationTokenBalanceChange,
+  SimulationTokenStandard,
+} from '@metamask/transaction-controller';
+import { useAsyncResultStrict } from '../../../hooks/useAsyncResult';
+import { getTokenStandardAndDetails } from '../../../store/actions';
+import { TokenStandard } from '../../../../shared/constants/transaction';
+import { Numeric } from '../../../../shared/modules/Numeric';
+import { getConversionRate } from '../../../ducks/metamask/metamask';
+import {
+  getConfirmationExchangeRates,
+  getTokenExchangeRates,
+} from '../../../selectors';
+import {
+  Amount,
+  BalanceChange,
+  FIAT_UNAVAILABLE,
+  NativeAssetIdentifier,
+  TokenAssetIdentifier,
+} from './types';
+
+const NATIVE_ASSET: NativeAssetIdentifier = { standard: TokenStandard.none };
+const NATIVE_DECIMALS = 18;
+const ERC20_DEFAULT_DECIMALS = 18;
+
+// Converts a SimulationTokenStandard to a TokenStandard
+const convertStandard = (standard: SimulationTokenStandard) => {
+  switch (standard) {
+    case SimulationTokenStandard.erc20:
+      return TokenStandard.ERC20;
+    case SimulationTokenStandard.erc721:
+      return TokenStandard.ERC721;
+    case SimulationTokenStandard.erc1155:
+      return TokenStandard.ERC1155;
+    default:
+      throw new Error(`Unknown token standard: ${standard}`);
+  }
+};
+
+// Calculates the asset amount based on the balance change and decimals
+const getAssetAmount = (
+  { isDecrease: isNegative, difference: quantity }: SimulationBalanceChange,
+  decimals: number,
+): Amount => {
+  const numeric = Numeric.from(quantity, 16)
+    .times(isNegative ? -1 : 1, 10)
+    .toBase(10)
+    .shiftedBy(decimals);
+  return { isNegative, quantity, decimals, numeric };
+};
+
+// Fetches token details for all the token addresses in the SimulationTokenBalanceChanges
+const fetchErc20Decimals = async (
+  addresses: Hex[],
+): Promise<Record<Hex, number>> => {
+  const uniqueAddresses = [...new Set(addresses)];
+  const tokenInfos = await Promise.all(
+    uniqueAddresses.map((address) => getTokenStandardAndDetails(address)),
+  );
+
+  return tokenInfos.reduce(
+    (result, { decimals }, index) => ({
+      ...result,
+      [uniqueAddresses[index]]: decimals
+        ? parseInt(decimals, 10)
+        : ERC20_DEFAULT_DECIMALS,
+    }),
+    {},
+  );
+};
+
+// Compiles the balance change for the native asset
+const getNativeBalanceChange = (
+  nativeBalanceChange: SimulationBalanceChange | undefined,
+  nativeFiatRate: number,
+): BalanceChange | undefined => {
+  if (!nativeBalanceChange) {
+    return undefined;
+  }
+  const asset = NATIVE_ASSET;
+  const amount = getAssetAmount(nativeBalanceChange, NATIVE_DECIMALS);
+  const fiatAmount = amount.numeric
+    .applyConversionRate(nativeFiatRate)
+    .toNumber();
+  return { asset, amount, fiatAmount };
+};
+
+// Compiles the balance changes for token assets
+const getTokenBalanceChanges = (
```

Comment:
Since this and the above are standard functions, should we use `function` syntax instead?